src/setup_payload/AdditionalDataPayloadGenerator.cpp

```diff
@@ -0,0 +1,122 @@
+    writer.Init(chip::System::PacketBufferHandle::New(chip::System::PacketBuffer::kMaxSize));
+
+    SuccessOrExit(err = writer.OpenContainer(AnonymousTag, kTLVType_Structure, innerWriter));
+
+    if (static_cast<int>(additionalDataFields & AdditionalDataFields::RotatingDeviceId) != 0)
+    {
+        // Generating Device Rotating Id
+        SuccessOrExit(err = generateRotatingDeviceId(lifetimeCounter, serialNumberBuffer, serialNumberBufferSize,
+                                                     rotatingDeviceIdBuffer, ArraySize(rotatingDeviceIdBuffer),
+                                                     rotatingDeviceIdBufferSize));
+
+        // Adding the rotating device id to the TLV data
+        SuccessOrExit(err = innerWriter.PutString(ContextTag(kRotatingDeviceIdTag), rotatingDeviceIdBuffer,
+                                                  static_cast<uint32_t>(rotatingDeviceIdBufferSize)));
+    }
+
+    SuccessOrExit(err = writer.CloseContainer(innerWriter));
+
+    SuccessOrExit(err = writer.Finalize(&bufferHandle));
+
+exit:
+    return err;
+}
+
+CHIP_ERROR AdditionalDataPayloadGenerator::generateRotatingDeviceId(uint16_t lifetimeCounter, const char * serialNumberBuffer,
+                                                                    size_t serialNumberBufferSize, char rotatingDeviceIdBuffer[],
+                                                                    size_t rotatingDeviceIdBufferSize,
+                                                                    size_t & rotatingDeviceIdValueOutputSize)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    Hash_SHA256_stream hash;
+    uint8_t outputBuffer[RotatingDeviceId::kRotatingDeviceIdMaxLength];
+    uint8_t hashOutputBuffer[kSHA256_Hash_Length];
+    BufferWriter outputBufferWriter(outputBuffer, ArraySize(outputBuffer));
+    size_t rotatingDeviceIdBufferIndex = 0;
+    uint8_t lifetimeCounterBuffer[2];
+
+    Put16(lifetimeCounterBuffer, lifetimeCounter);
+
+    VerifyOrExit(rotatingDeviceIdBufferSize >= RotatingDeviceId::kRotatingDeviceIdHexMaxLength, err = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    // Computing the Rotating Device Id
+    // RDI = Lifetime_Counter + SuffixBytes(SHA256(Serial_Number + Lifetime_Counter), 16)
+
+    SuccessOrExit(err = hash.Begin());
+    SuccessOrExit(err = hash.AddData(Uint8::from_const_char(serialNumberBuffer), serialNumberBufferSize));
+    SuccessOrExit(err = hash.AddData(lifetimeCounterBuffer, sizeof(lifetimeCounter)));
+    SuccessOrExit(err = hash.Finish(hashOutputBuffer));
+
+    outputBufferWriter.Put16(lifetimeCounter);
+    outputBufferWriter.Put(&hashOutputBuffer[kSHA256_Hash_Length - RotatingDeviceId::kRotatingDeviceIdHashSuffixLength],
+                           RotatingDeviceId::kRotatingDeviceIdHashSuffixLength);
+
+    for (; rotatingDeviceIdBufferIndex < outputBufferWriter.Needed(); rotatingDeviceIdBufferIndex++)
+    {
+        snprintf(&rotatingDeviceIdBuffer[rotatingDeviceIdBufferIndex * 2],
```

Comment:
Could we have some overflow tests when rotatingDeviceIdBufferSize is too small?

Reply:
i already opened an issue to add unit tests for the generation logic #4825

Reply:
This is the reverse of how it should be done. The unit tests should come as part of the original code, otherwise how are we to know that it works as expected in this PR, and how it was tested? This code is landing in a way that can be enabled, but without the tests.

Since this is a PR that brings in:

- Data that appears on the air
- Data which may be read by external radio devices over BLE or IP bearing network
- Computation of complex payloads over multiple buffers in multiple formats (TLV, hex strings, etc)

It meets the bar for needing tests as part of the original PR. Bugs in this code can leak memory on the air.